File: widgetssdk/src/test/java/com/glia/widgets/core/engagement/domain/MapOperatorUseCaseTest.java
package com.glia.widgets.core.engagement.domain;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import com.glia.androidsdk.Operator;
import com.glia.androidsdk.chat.ChatMessage;
import com.glia.androidsdk.chat.OperatorMessage;
import com.glia.widgets.core.model.TestOperator;

import org.junit.Before;
import org.junit.Test;

import java.util.Optional;

import io.reactivex.Single;

public class MapOperatorUseCaseTest {
    MapOperatorUseCase mapOperatorUseCase;
    GetOperatorUseCase getOperatorUseCase;
    Operator operator;

    @Before
    public void setUp() throws Exception {
        getOperatorUseCase = mock(GetOperatorUseCase.class);
        mapOperatorUseCase = new MapOperatorUseCase(getOperatorUseCase);
        operator = TestOperator.DEFAULT;
    }

    @Test
    public void execute_returnsInternalMessageWithoutOperator_whenChatMessageIsVisitor() {
        ChatMessage chatMessage = mock(ChatMessage.class);
        when(chatMessage.isOperator()).thenReturn(false);

        mapOperatorUseCase.execute(chatMessage)
                .test()
                .assertComplete()
                .assertNever(messageInternal -> messageInternal.getOperator().isPresent());
    }

    @Test
    public void execute_returnsInternalMessageWithOperator_whenChatMessageIsOperator() {
        OperatorMessage chatMessage = mock(OperatorMessage.class);
        when(chatMessage.isOperator()).thenReturn(true);
        when(chatMessage.getOperatorId()).thenReturn(operator.getId());

        when(getOperatorUseCase.execute(operator.getId())).thenReturn(Single.just(Optional.of(operator)));

        mapOperatorUseCase.execute(chatMessage)
                .doOnSuccess(messageInternal -> assertEquals(messageInternal.getOperatorId().get(), operator.getId()))
                .test()
                .assertComplete()
                .assertNever(messageInternal -> !messageInternal.getOperator().isPresent());
    }
}

File: widgetssdk/src/test/java/com/glia/widgets/core/engagement/domain/MapOperatorUseCaseTest.kt
package com.glia.widgets.core.engagement.domain

import com.glia.androidsdk.Operator
import com.glia.androidsdk.chat.Chat
import com.glia.androidsdk.chat.ChatMessage
import com.glia.androidsdk.chat.OperatorMessage
import com.glia.widgets.core.engagement.domain.model.ChatMessageInternal
import com.glia.widgets.core.model.TestOperator
import io.reactivex.Single
import junit.framework.TestCase.assertEquals
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import java.util.Optional
import kotlin.properties.Delegates

class MapOperatorUseCaseTest {
    private var mapOperatorUseCase: MapOperatorUseCase by Delegates.notNull()
    var getOperatorUseCase: GetOperatorUseCase by Delegates.notNull()
    var operator: Operator by Delegates.notNull()

    @Before
    fun setUp() {
        getOperatorUseCase = mock()
        mapOperatorUseCase = MapOperatorUseCase(getOperatorUseCase)
        operator = TestOperator.DEFAULT
    }

    @Test
    fun execute_returnsInternalMessageWithoutOperator_whenChatMessageIsVisitor() {
        val chatMessage: ChatMessage = mock()
        whenever(chatMessage.senderType) doReturn Chat.Participant.SYSTEM
        mapOperatorUseCase(chatMessage)
            .test()
            .assertComplete()
            .assertNever { it.operator.isPresent }
    }

    @Test
    fun execute_returnsInternalMessageWithOperator_whenChatMessageIsOperator() {
        val chatMessage: OperatorMessage = mock()
        whenever(chatMessage.senderType) doReturn Chat.Participant.OPERATOR
        whenever(chatMessage.operatorId) doReturn operator.id
        whenever(getOperatorUseCase.execute(operator.id)) doReturn (Single.just(Optional.of(operator)))
        mapOperatorUseCase(chatMessage)
            .doOnSuccess { assertEquals(it.operatorId.get(), operator.id) }
            .test()
            .assertComplete()
            .assertNever { !it.operator.isPresent }
    }
}